File: src/core/context-builder.ts
import type {
  AgentContext,
  AgentMessage,
  ModelRequest,
} from "./types.js";

/**
 * ContextBuilder 决定“完整历史中的哪些内容在本轮发给模型”。
 *
 * ConversationStore 仍然保存完整历史；ContextBuilder 只创建一次模型请求的只读快照。
 * 这样截断不会悄悄删除用户的 memory，也不会污染 AgentContext。
 */
export type BuiltContext = Pick<
  ModelRequest,
  "systemPrompt" | "messages" | "tools"
>;

export type ContextBuilder = {
  build(context: Readonly<AgentContext>): Promise<BuiltContext> | BuiltContext;
};

/** Tokenizer 属于模型协议；注入后，ContextBuilder 不需要猜测供应商的切词规则。 */
export type TokenCounter = {
  count(text: string): number;
};

export type SummaryProvider = {
  summarize(messages: readonly AgentMessage[]): Promise<string> | string;
};

export type RecentContextBuilderOptions = {
  maxMessages?: number;
  maxCharacters?: number;
};

/**
 * 保留最近完整对话轮次的教学版 ContextBuilder。
 *
 * 它使用字符数近似 token 数，避免为了教学引入 tokenizer 依赖。真正的生产系统可以
 * 实现同一个 ContextBuilder 接口，换成精确 token 预算、摘要或检索结果。
 */
export class RecentContextBuilder implements ContextBuilder {
  private readonly maxMessages: number;
  private readonly maxCharacters: number;

  constructor(options: RecentContextBuilderOptions = {}) {
    this.maxMessages = options.maxMessages ?? 40;
    this.maxCharacters = options.maxCharacters ?? 50_000;
    assertPositiveInteger("maxMessages", this.maxMessages);
    assertPositiveInteger("maxCharacters", this.maxCharacters);
  }

  build(context: Readonly<AgentContext>): BuiltContext {
    return {
      systemPrompt: context.systemPrompt,
      messages: selectRecentTurns(
        context.messages,
        this.maxMessages,
        this.maxCharacters,
      ),
      tools: context.tools,
    };
  }
}

export type TokenContextBuilderOptions = {
  maxTokens: number;
  tokenCounter: TokenCounter;
  summarizer?: SummaryProvider;
};

/**
 * 按调用者提供的真实 tokenizer 保留完整轮次，并可把被裁掉的旧轮次压成摘要。
 * 摘要只进入本轮 system prompt；ConversationStore 中的原始消息不会被覆盖。
 */
export class TokenContextBuilder implements ContextBuilder {
  constructor(private readonly options: TokenContextBuilderOptions) {
    assertPositiveInteger("maxTokens", options.maxTokens);
  }

  async build(context: Readonly<AgentContext>): Promise<BuiltContext> {
    const turns = groupIntoTurns(context.messages);
    const selected: AgentMessage[][] = [];
    let used =
      this.options.tokenCounter.count(context.systemPrompt) +
      this.options.tokenCounter.count(JSON.stringify(context.tools));
    for (let index = turns.length - 1; index >= 0; index -= 1) {
      const turn = turns[index];
      if (!turn) continue;
      const cost = this.options.tokenCounter.count(JSON.stringify(turn));
      if (selected.length > 0 && used + cost > this.options.maxTokens) break;
      selected.unshift(turn);
      used += cost;
    }
    const selectedCount = selected.reduce(
      (total, turn) => total + turn.length,
      0,
    );
    const omitted = context.messages.slice(
      0,
      context.messages.length - selectedCount,
    );
    const rawSummary = omitted.length > 0 && this.options.summarizer
      ? (await this.options.summarizer.summarize(omitted)).trim()
      : "";
    const summaryWrapper = "\n\n<conversation_summary>\n\n</conversation_summary>";
    const summaryBudget = Math.max(
      0,
      this.options.maxTokens -
        used -
        this.options.tokenCounter.count(summaryWrapper),
    );
    const summary = fitTextToTokenBudget(
      rawSummary,
      summaryBudget,
      this.options.tokenCounter,
    );
    return {
      systemPrompt: summary
        ? `${context.systemPrompt}\n\n<conversation_summary>\n${summary}\n</conversation_summary>`
        : context.systemPrompt,
      messages: selected.flat(),
      tools: context.tools,
    };
  }
}

function fitTextToTokenBudget(
  text: string,
  maxTokens: number,
  counter: TokenCounter,
): string {
  if (!text || maxTokens <= 0) return "";
  let low = 0;
  let high = text.length;
  while (low < high) {
    const middle = Math.ceil((low + high) / 2);
    if (counter.count(text.slice(0, middle)) <= maxTokens) low = middle;
    else high = middle - 1;
  }
  return text.slice(0, low);
}

/** 无模型、可预测的教学摘要器；生产环境可在同一接口接专用摘要模型。 */
export class ExtractiveSummaryProvider implements SummaryProvider {
  constructor(private readonly maxCharacters = 2_000) {
    assertPositiveInteger("maxCharacters", maxCharacters);
  }

  summarize(messages: readonly AgentMessage[]): string {
    const lines = messages.map((message) => {
      if (message.role === "user") {
        const content = typeof message.content === "string"
          ? message.content
          : message.content.map((block) =>
              block.type === "text" ? block.text : `[image: ${block.source.mediaType}]`
            ).join(" ");
        return `user: ${content}`;
      }
      if (message.role === "tool") {
        return `tool(${message.toolName}): ${message.content}`;
      }
      const text = message.content
        .filter((block) => block.type === "text")
        .map((block) => block.text)
        .join(" ");
      return `assistant: ${text || `[${message.stopReason}]`}`;
    });
    return lines.join("\n").slice(-this.maxCharacters);
  }
}

function selectRecentTurns(
  messages: readonly AgentMessage[],
  maxMessages: number,
  maxCharacters: number,
): AgentMessage[] {
  const turns = groupIntoTurns(messages);
  const selected: AgentMessage[][] = [];
  let messageCount = 0;
  let characterCount = 0;

  // 从最新轮次向前装填。即使最新轮本身超过预算，也必须完整保留，避免只留下
  // tool result，却丢掉对应的 assistant tool call。
  for (let index = turns.length - 1; index >= 0; index -= 1) {
    const turn = turns[index];
    if (!turn) continue;
    const nextMessageCount = messageCount + turn.length;
    const nextCharacterCount =
      characterCount + turn.reduce(countMessageCharacters, 0);
    const exceedsBudget =
      nextMessageCount > maxMessages ||
      nextCharacterCount > maxCharacters;
    if (selected.length > 0 && exceedsBudget) break;
    selected.unshift(turn);
    messageCount = nextMessageCount;
    characterCount = nextCharacterCount;
  }

  return selected.flat();
}

/**
 * 一条 user message 开始一个新轮次；之后的 assistant/tool messages 属于同一轮。
 */
export function groupIntoTurns(
  messages: readonly AgentMessage[],
): AgentMessage[][] {
  const turns: AgentMessage[][] = [];
  for (const message of messages) {
    if (message.role === "user" || turns.length === 0) turns.push([]);
    turns.at(-1)?.push(message);
  }
  return turns;
}

function countMessageCharacters(
  total: number,
  message: AgentMessage,
): number {
  return total + JSON.stringify(message).length;
}

function assertPositiveInteger(name: string, value: number): void {
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error(`${name} must be a positive integer.`);
  }
}

File: python/from_scratch_agent/context_builder.py
"""ContextBuilder：按完整轮次裁剪，并可摘要旧历史。"""

import json
from html import escape
from dataclasses import dataclass
from typing import Protocol

from .types import AgentContext, Message
from .memory_index import MemoryIndex, MemoryKind


class TokenCounter(Protocol):
    def count(self, text: str) -> int: ...


class SummaryProvider(Protocol):
    def summarize(self, messages: list[Message]) -> str: ...


class ContextBuilder(Protocol):
    def build(self, context: AgentContext) -> AgentContext: ...


@dataclass
class TokenContextBuilder:
    """使用注入的 provider tokenizer 计算预算，不猜测模型切词规则。"""

    max_tokens: int
    token_counter: TokenCounter
    summarizer: SummaryProvider | None = None

    def build(self, context: AgentContext) -> AgentContext:
        if self.max_tokens <= 0:
            raise ValueError("max_tokens 必须为正整数")
        turns = _group_turns(context.messages)
        selected: list[list[Message]] = []
        tool_payload = [
            {
                "name": tool.name,
                "description": tool.description,
                "input_schema": tool.input_schema,
            }
            for tool in context.tools
        ]
        used = self.token_counter.count(context.system_prompt) + self.token_counter.count(
            json.dumps(tool_payload, ensure_ascii=False, separators=(",", ":"))
        )
        for turn in reversed(turns):
            cost = self.token_counter.count(
                json.dumps(turn, ensure_ascii=False, separators=(",", ":"))
            )
            if selected and used + cost > self.max_tokens:
                break
            selected.insert(0, turn)
            used += cost
        kept = sum(len(turn) for turn in selected)
        omitted = context.messages[: len(context.messages) - kept]
        raw_summary = (
            self.summarizer.summarize(omitted).strip()
            if omitted and self.summarizer
            else ""
        )
        wrapper = "\n\n<conversation_summary>\n\n</conversation_summary>"
        summary_budget = max(
            0,
            self.max_tokens - used - self.token_counter.count(wrapper),
        )
        summary = _fit_text(raw_summary, summary_budget, self.token_counter)
        prompt = context.system_prompt
        if summary:
            prompt += f"\n\n<conversation_summary>\n{summary}\n</conversation_summary>"
        return AgentContext(
            system_prompt=prompt,
            messages=[message for turn in selected for message in turn],
            tools=context.tools,
        )


@dataclass
class ExtractiveSummaryProvider:
    max_characters: int = 2000

    def summarize(self, messages: list[Message]) -> str:
        lines = []
        for message in messages:
            role = message.get("role", "unknown")
            content = message.get("content", "")
            if isinstance(content, list):
                content = " ".join(
                    str(block.get("text", ""))
                    for block in content
                    if isinstance(block, dict) and block.get("type") == "text"
                )
            lines.append(f"{role}: {content}")
        return "\n".join(lines)[-self.max_characters :]


@dataclass
class MemoryRecallContextBuilder:
    index: MemoryIndex
    delegate: ContextBuilder | None = None
    limit: int = 5
    kinds: list[MemoryKind] | None = None

    def build(self, context: AgentContext) -> AgentContext:
        built = self.delegate.build(context) if self.delegate else AgentContext(
            context.system_prompt, list(context.messages), context.tools
        )
        query = next(
            (str(message.get("content", "")) for message in reversed(context.messages)
             if message.get("role") == "user"),
            "",
        )
        memories = self.index.search(query, self.limit, self.kinds) if query else []
        if memories:
            section = "\n".join(
                f'<memory kind="{item.kind}" id="{escape(item.id)}">'
                f'{escape(item.content)}</memory>'
                for item in memories
            )
            built.system_prompt += f"\n\n# Relevant memories\n\n{section}"
        return built


def _group_turns(messages: list[Message]) -> list[list[Message]]:
    turns: list[list[Message]] = []
    for message in messages:
        if message.get("role") == "user" or not turns:
            turns.append([])
        turns[-1].append(message)
    return turns


def _fit_text(text: str, max_tokens: int, counter: TokenCounter) -> str:
    if not text or max_tokens <= 0:
        return ""
    low, high = 0, len(text)
    while low < high:
        middle = (low + high + 1) // 2
        if counter.count(text[:middle]) <= max_tokens:
            low = middle
        else:
            high = middle - 1
    return text[:low]
